addons/mail/static/src/models/message/message.js

```diff
@@ -445,9 +510,158 @@ function factory(dependencies) {
+             * We remove the data-message-collapse attr when the processing is done.
+             */
+            let collapsable = contentAsHMTL.querySelector("[data-message-collapse]");
+            while (collapsable) {
+
+                collapsable.removeAttribute("data-message-collapse");
+                
+                /**
+                 * We create a hash based on the content of the node. 
+                 * We need to do this because we have no other way to compare nodes at each different update.
+                 * We use the message id as prefix to make sure not to modify other messages that could have the same content.
+                 * It will be kept in a map to keep track of the opened/closed state of the collapsable contents.
+                 */
+                const hash = this.id + '' + collapsable.innerHTML.split('').reduce((a, b) => { a = ((a << 5) - a) + b.charCodeAt(0); return a & a }, 0);
+                
+                /**
+                 * If the map keeping track of the opened collapsable contents doesn't contain the current hash, it means it's still closed. 
+                 * So we replace the content with a Read More possibility.
+                 * Otherwise, we keep the content and add to it a Read Less possibility.
+                 */
+                if (!this.currentlyOpenedReadMoreHashes.includes(hash)) {
+                    collapsable.outerHTML = `<span class="o_Message_contentCollapsed">Content collapsed...</span><span class="o_Message_readMoreLess" data-read-more-hash="${hash}">   Read More</span>`
+                } 
+                else {
+                    collapsable.outerHTML = collapsable.outerHTML + `<span class="o_Message_readMoreLess" data-read-less-hash="${hash}">   Read Less</span>`
+                }
+
+                collapsable = contentAsHMTL.querySelector("[data-message-collapse]");
+
+            }
+
+            /**
+             * We used a DOM Element, but all the process use DOM as string, so we get
+             * back the inner html.
+             */
+            return contentAsHMTL.innerHTML;
+        }
+        
+        /**
+         * Add a hash to the map keeping track of open collapsable contents.
+         * Basically means it opens it.
+         * 
+         * @param {String} hash 
+         */
+        readMoreElementClicked(hash) {
+            this.currentlyOpenedReadMoreHashes.push(hash);
+        }
+
+        /**
+         * Remove a hash to the map keeping track of open collapsable contents.
+         * Basically means it closes it.
+         * 
+         * @param {String} hash 
+         */
+        readLessElementClicked(hash) {
+            for (let i = 0; i < this.currentlyOpenedReadMoreHashes.length; i++) { 
+                if (this.currentlyOpenedReadMoreHashes[i] === hash) { 
+                    this.currentlyOpenedReadMoreHashes.splice(i, 1); 
+                    i--; 
+                }
+            }
+        }
```

Comment:
This really makes it looks like you should be using a Set instead of an array.

Reply:
Or even better a new model and an x2many relation instead of an array? Relations are automatically Set and properly trigger render on update without doing array tricks.

Reply:
@seb-odoo Isn't a model for just a set of values a bit too much. That would create a new file that with a somewhat obscur name. Doesn't seem like it's a good idea in this case imo.